resolwe/observers/signals.py

```diff
@@ -0,0 +1,49 @@
+"""ORM signal handlers."""
+
+from django import dispatch
+from django.db.models import signals as model_signals
+
+from .models import Observer
+from .protocol import CHANGE_TYPE_DELETE, CHANGE_TYPE_UPDATE
+
+# Global 'in migrations' flag to ignore signals during migrations.
+# Signal handlers that access the database can crash the migration process.
+IN_MIGRATIONS = False
+
+
+@dispatch.receiver(model_signals.pre_migrate)
+def model_pre_migrate(*args, **kwargs):
+    """Set 'in migrations' flag."""
+    global IN_MIGRATIONS
+    IN_MIGRATIONS = True
+
+
+@dispatch.receiver(model_signals.post_migrate)
+def model_post_migrate(*args, **kwargs):
+    """Clear 'in migrations' flag."""
+    global IN_MIGRATIONS
+    IN_MIGRATIONS = False
+
+
+@dispatch.receiver(model_signals.post_save)
+def observe_model_modification(sender, instance, created=False, **kwargs):
+    """Receive model updates."""
+    global IN_MIGRATIONS
+    if IN_MIGRATIONS:
+        return
+
+    # Create signals will be caught when the PermissionModel is added.
+    if created:
+        return
+
+    Observer.observe_instance_changes(instance, CHANGE_TYPE_UPDATE)
+
+
+@dispatch.receiver(model_signals.pre_delete)
+def observe_model_deletion(sender, instance, **kwargs):
```

Comment:
Is it possible that objects is not really deleted after pre_delete is triggered?

Reply:
Yes, if the transaction is rolled back. That's why the actual notification is sent through `transaction.on_commit`, but that code lies two frames deeper, in the Subscription model.
https://github.com/genialis/resolwe/blob/23b4313e97b6e88d5a38a8dcfa0633ab0a0b7e5d/resolwe/observers/models.py#L176
The reason is that we're trying to collect information about the to-be-deleted object before it's actually deleted.